AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLDocument/SingleDirectiveGraphQLDocument.swift

```diff
@@ -0,0 +1,111 @@
+/// Therefore, documents represented by concrete implementations provide a single GraphQL
+/// operation based on a defined `Model.Type`.
+public protocol SingleDirectiveGraphQLDocument {
+    /// The `GraphQLOperationType` a concrete implementation represents the
+    /// GraphQL operation of the document
+    var operationType: GraphQLOperationType { get set }
+
+    /// The name of the document. This is useful to inspect the response, since it will
+    /// contain the name of the document as the key to the response value.
+    var name: String { get set }
+
+    /// Input parameters and its values on the GraphQL document
+    var inputs: [GraphQLParameterName: GraphQLDocumentInput] { get set }
+
+    /// The selection set of the document, used to specify the response data returned by the service.
+    var selectionSet: SelectionSet? { get set }
+
+    /// Simple constructor to be implemented by the concrete types, used by the `copy` method.
+    init(operationType: GraphQLOperationType,
+         name: String,
+         inputs: [GraphQLParameterName: GraphQLDocumentInput],
+         selectionSet: SelectionSet?)
+}
+
+// Provides default implementation
+extension SingleDirectiveGraphQLDocument {
+
+    /// Method to create a deep copy of the document, useful for `SingleDirectiveGraphQLDocumentDecorator` decorators
+    /// when decorating a document and returning a new document.
+    public func copy(operationType: GraphQLOperationType? = nil,
+                     name: String? = nil,
+                     inputs: [GraphQLParameterName: GraphQLDocumentInput]? = nil,
+                     selectionSet: SelectionSet? = nil) -> Self {
+
+        return Self.init(operationType: operationType ?? self.operationType,
+                         name: name ?? self.name,
+                         inputs: inputs ?? self.inputs,
+                         selectionSet: selectionSet ?? self.selectionSet)
+    }
+
+    /// Returns an empty `variables` object when there are no `inputs` required, Otherwise, consolidates the `inputs`
+    /// into a single object that can be used for the GraphQL request.
+    public var variables: [String: Any] {
+        var variables = [String: Any]()
+        inputs.forEach { input in
+            switch input.value.value {
+            case .object(let values):
+                variables.updateValue(values, forKey: input.key)
+            case .value(let value):
+                variables.updateValue(value, forKey: input.key)
+            }
+
+        }
+
+        return variables
+    }
+
+    /// Provides default construction of the graphQL document based on the components of the document.
+    public var stringValue: String {
+
+        var selectionSetString = ""
+        if let selectionSet = selectionSet {
+            selectionSetString = selectionSet.fields.map { $0.toString() }.joined(separator: "\n    ")
+        }
+
+        if !inputs.isEmpty {
```

Comment:
nit: change this to a `guard !inputs.isEmpty else return { your empty-case value below }`